SmartPointers/shared_pointer.hpp

```diff
@@ -1,80 +1,173 @@
-//
-//  shared_pointer.hpp
-//  Pointer
-//
-//  Created by Srushti Desai on 12/27/23.
-//
-
-#ifndef shared_pointer_hpp
-#define shared_pointer_hpp
-
-using namespace std;
-#include <iostream>
-template<class T>
-class SharedPointer
-{
-private:
-    T* m_ptr;
-    unsigned int* m_count;
-public:
-     SharedPointer():m_ptr(nullptr), m_count(new unsigned int(0))
+
+
+
+
+template<typename T>
+class SharedPointer{
+
+    public:
+
+    //Constructor
+    SharedPointer():m_ptr(nullptr), count(new int(1))
     {
-        
+
     }
 
-    ~SharedPointer(){
-        //cleanup();
-        
+    //Paramatized Constructor
+     SharedPointer(T *ptr):m_ptr(ptr), count(new int(1))
+    {
+
     }
 
-    SharedPointer(T * ptr) : m_ptr(ptr), m_count(new unsigned int(1))
-        {
-        }
+    //Copy Constructor
+    SharedPointer(const SharedPointer& p): m_ptr(p.m_ptr),
+    count(p.count)
+    {
+        (*count)++;
+    }
+
+    //Copy assigment operator
+    SharedPointer& operator=(const SharedPointer& p )
+    {
+         if (this != &p) {
+            *count = *count> 0 ? *count-1: *count;
```

Comment:
Why are you checking it it could be zero already? Is it possible ? Are we maintaining any invariant to prevent that?

Reply:
This is to make sure in case